gap/oper.gi

```diff
@@ -377,6 +377,237 @@ InstallMethod(DigraphClosure,
 [IsImmutableDigraph, IsPosInt],
 {D, k} -> MakeImmutable(DigraphClosure(DigraphMutableCopy(D), k)));
 
+InstallMethod(DigraphContractEdge,
+"for a mutable digraph and two positive integers",
+[IsMutableDigraph, IsPosInt, IsPosInt],
+function(D, u, v)
+  local w, neighbours, transformations_to_edge, transformations_to_old_edges,
+  old_edge, new_edge, neighbour, t, copy, vertices;
+
+  if not IsDigraphEdge(D, u, v) then  # Check if [u, v] is an edge in digraph D
+    ErrorNoReturn("u, v is not an edge of D");
+  fi;
+
+  if IsMultiDigraph(D) then  # multi digraphs cannot have their edges contracted
+    ErrorNoReturn("Cannot contract an edge for a MultiDigraph");
+  fi;
+
+  if u = v then  # edge should not be contracted if u = v
+    ErrorNoReturn("Cannot contract an edge with the same source and range");
+  fi;
+
+  if IsDigraphEdge(D, v, u) then  # if (v, u) is an edge,
+    DigraphRemoveEdge(D, v, u);   # disallow loops, so also remove (v, u)
+  fi;
+
+  # remove the contracted edge
+  DigraphRemoveEdge(D, u, v);
+
+  copy := DigraphImmutableCopy(D);  # create a copy of the original digraph,
```

Comment:
If `D` doesn't have edge or vertex labels, then this copy is unnecessary, we should just return `D` right away here without doing anything extra. 

Also: even if `D` does have edge and/or vertex labels, then we don't really require the copy of `D` but only of its edge and/or vertex labels, and even then probably only the labels of the vertices or edge incident to the edge being collapsed. I think this should be updated to copy only the very least amount of data from `D` required.